File: src/javascript/valid_parenthesis.js
function isValidParenthesis(s) {
  if (s.length % 2 === 1) return false;

  let store = [];
  let tokens = { "(": ")", "{": "}", "[": "]" };

  for (const item of s.split("")) {
    if (tokens[item]) {
      store.push(item);
    } else {
      let lastItem = store.pop();
      if (tokens[lastItem] !== item) {
        return false;
      }
    }
  }

  return store.length === 0;
}

File: src/go/valid_parenthesis.go
package main

import "fmt"

func IsValidParenthesis(str string) bool {
	if len(str)%2 == 1 || len(str) == 0 {
		return false
	}

	store := []string{}
	tokens := make(map[string]string)
	tokens["("] = ")"
	tokens["{"] = "}"
	tokens["["] = "]"

	for _, char := range str {
		if tokens[string(char)] != "" {
			store = append(store, string(char))
		} else {
			if len(store)-1 < 0 {
				return false
			}

			lastChar := store[len(store)-1]

			if tokens[lastChar] == string(char) {
				store = store[:len(store)-1]
			} else {
				return false
			}
		}
	}

	return len(store) == 0
}

func main() {
	input := "){"
	ans := IsValidParenthesis(input)

	fmt.Println(ans)
}
